homeassistant/components/qbus/entity.py

```diff
@@ -0,0 +1,101 @@
+"""Base class for Qbus entities."""
+
+from abc import ABC, abstractmethod
+import re
+
+from qbusmqttapi.discovery import QbusMqttOutput
+from qbusmqttapi.factory import QbusMqttMessageFactory, QbusMqttTopicFactory
+from qbusmqttapi.state import QbusMqttState
+
+from homeassistant.components.mqtt import ReceiveMessage, client as mqtt
+from homeassistant.helpers.device_registry import DeviceInfo, format_mac
+from homeassistant.helpers.entity import Entity
+
+from .const import DOMAIN, MANUFACTURER
+
+_REFID_REGEX = re.compile(r"^\d+\/(\d+(?:\/\d+)?)$")
+
+
+def format_ref_id(ref_id: str) -> str | None:
+    """Format the Qbus ref_id."""
+    matches: list[str] = re.findall(_REFID_REGEX, ref_id)
+
+    if len(matches) > 0:
+        if ref_id := matches[0]:
+            return ref_id.replace("/", "-")
+
+    return None
+
+
+class QbusEntity(Entity, ABC):
+    """Representation of a Qbus entity."""
+
+    _attr_has_entity_name = True
+    _attr_should_poll = False
+
+    def __init__(
+        self,
+        mqtt_output: QbusMqttOutput,
+        entity_id_format: str,
+        *,
+        id_suffix: str | None = None,
+    ) -> None:
+        """Initialize the Qbus entity."""
+
+        self._topic_factory = QbusMqttTopicFactory()
+        self._message_factory = QbusMqttMessageFactory()
+
+        ref_id = format_ref_id(mqtt_output.ref_id)
+
+        id_suffix = id_suffix or ""
+        self._attr_unique_id = (
+            f"ctd_{mqtt_output.device.serial_number}_{ref_id}{id_suffix}"
+        )
+        self.entity_id = entity_id_format.format(
+            f"qbus_{mqtt_output.device.serial_number}_{ref_id}{id_suffix}"
+        )
+
+        self._attr_device_info = DeviceInfo(
+            name=mqtt_output.name.title(),
+            manufacturer=MANUFACTURER,
+            identifiers={(DOMAIN, f"{mqtt_output.device.serial_number}_{ref_id}")},
+            suggested_area=mqtt_output.location.title(),
+            via_device=(DOMAIN, format_mac(mqtt_output.device.mac)),
+        )
+
+        self._attr_extra_state_attributes = {
+            "controller_id": mqtt_output.device.id,
```

Comment:
if you use `via_device`, is `controller_id` needed as a state attribute?

Reply:
The Qbus MQTT gateway assigns ID's to controllers and outputs, but those ID's could change when making changes to the Qbus system. 

Having this as an attribute is useful for creating automations that trigger based on receiving Qbus MQTT messages, for instance Qbus scene events. The topic contains the controller and output ID.

It can also help in diagnosing problems.